### ISSUE TITLE
Uuid referenced not found in database

### ISSUE THREAD

**Issue by tumbl3w33d** [kanidm/kanidm#1904]
### I did this

I am trying to update group attributes via API. My python code (which I can provide, but I don't think it's relevant) shows the following debug output:

```
# a simple get
requesting groups via GET from https://kanidm-playground.mylocal:8443/v1/group

# the response, reduced to the relevant group that I want to update
api returned 200 - b'[[…],{"attrs":{"class":["group","object"],"name":["bookstack_admins"],"spn":["bookstack_admins@kanidm-playground.mylocal"],"uuid":["2e9081e4-424b-4f43-8a56-be7be8c5c147"]}},[…]]'

# kind of duplicate output, because I locally check if any attribute needs an update
retrieved group {'attrs': {'class': ['group', 'object'], 'name': ['bookstack_admins'], 'spn': ['bookstack_admins@kanidm-playground.mylocal'], 'uuid': ['2e9081e4-424b-4f43-8a56-be7be8c5c147']}}

# trying to set the member attr via PUT request
sending payload ['robot.bookstack'] by PUT to https://kanidm-playground.mylocal:8443/v1/group/2e9081e4-424b-4f43-8a56-be7be8c5c147/_attr/member
500 - b'{"plugin":{"referentialintegrity":"Uuid referenced not found in database"}}'
```

### I expected the following

The group should have the member added. I also would expect a 404 in case of unknown uuid. This looks like an unhandled error.

### Kanidm version (and git commit)

```bash
docker exec -it kanidm kanidmd version
kanidmd 1.1.0-alpha.12
```

### Operating System / Version

`Linux milkshake 6.1.38-1-MANJARO #1 SMP PREEMPT_DYNAMIC Wed Jul  5 23:49:30 UTC 2023 x86_64 GNU/Linux`

Kanidm is running in the official container.

### Any other comments

I did exactly the same procedure with person objects and did not encounter this issue. Btw, I could't quite figure out the difference between PUT and POST yet. Saw something about POST being for appending, but would prefer the explanation of your idea behind it.

**Comment by Firstyear** [comment]
Yeah, I think a better error here is a valid request. Refint is saying "something you asked for isn't present" but it should say what it was. I think the problem is that by that point we have stripped the actual data itself because the name2uuid step assigns a dummy uuid that cannever and will never exist, so we actually lose the data we need to report.

So I think we need a broader rethink and improvement of name/spn/uuid here and the interplay between them because we need better responses to errors like this.

**Comment by Firstyear** [comment]
Oh and yes, 404 would have been nicer, but also that goes into our error handling that is too coarse to what code it returns. That's also a bigger issue I've had in mind that we need finer error types that have better status codes when returned up to the http layers.

**Comment by Seba-T** [comment]
> we need finer error types that have better status codes when returned up to the http layers. 

Don’t we already have some ‘into’ impl for errors into https status codes? Wouldn’t it be just a matter of fine tuning that specific section?

**Comment by Firstyear** [comment]
@Seba-T The problem is that right now sometimes we want it as a 404, sometimes we want it to be a 200, sometimes we want it to be a 500. Because errors are too coarse right now relative to the operation type, we are picking the wrong type here.

At the same time, I also somewhat disagree with @yaleman on this one that sometimes I think we should return a 200 and then an encoded result in the json, since after all the operation did "successfully process" even though it's inner result of the operation is an error. But I can also see the other side that sometimes the specific result codes are good for developers for scripting etc. Saying this those result codes in http often aren't precise enough to truly convey they meaning of our operation status so we have to botch them a bit like this situation.

**Comment by yaleman** [comment]
> At the same time, I also somewhat disagree with @yaleman on this one that sometimes I think we should return a 200 and then an encoded result in the json, since after all the operation did "successfully process" even though it's inner result of the operation is an error. 

I'm open to any option as long as it's ergonomically easy for consumers 😄 A 500 here's definitely a bork, for sure.

**Comment by Firstyear** [comment]
I think we need a bigger wholistic effort to improve fault/error reporting. We can resolve this I think by having a custom error for this situation to make a 404 in the refint handler, but we also should consider a better error manager. I've been thinking about solaris fault management and how it had unique error codes and reporting actions on what the admin needed to do for every error, and we should consider and think about that here I think for Kanidm. Because I think this also will tie in with how we deliver errors to the client and in http status codes to help distinguish what is an error as in the server is broke, or what's an error as in "you the user need to change a constraint or input".

**Comment by Firstyear** [comment]
![image](https://github.com/kanidm/kanidm/assets/271005/edcdec97-998d-4897-bdeb-7335c5fdfe44)

**Comment by yaleman** [comment]
I went down a rabbit hole on our existing event ids the other night and realised they were just enum values leaking, I could definitely see a use for defined errors and a path to find solutions 😄

**Comment by yaleman** [comment]
I'm pulling this out of the bug line because it's less a bug and more a "terrible response that needs to be improved.
